test/e2e/specs/editor/various/undo.spec.js

```diff
@@ -0,0 +1,501 @@
+/**
+ * WordPress dependencies
+ */
+const { test, expect } = require( '@wordpress/e2e-test-utils-playwright' );
+
+test.use( {
+	undoUtils: async ( { page }, use ) => {
+		await use( new UndoUtils( { page } ) );
+	},
+} );
+
+class UndoUtils {
+	constructor( { page } ) {
+		this.page = page;
+
+		this.getSelection = this.getSelection.bind( this );
+	}
+
+	async getSelection() {
+		return await this.page.evaluate( () => {
+			const selectedBlockId = window.wp.data
+				.select( 'core/block-editor' )
+				.getSelectedBlockClientId();
+			const blockIndex = window.wp.data
+				.select( 'core/block-editor' )
+				.getBlockIndex( selectedBlockId );
+
+			if ( blockIndex === -1 ) {
+				return {};
+			}
+
+			return {
+				blockIndex,
+				startOffset: window.wp.data
+					.select( 'core/block-editor' )
+					.getSelectionStart().offset,
+				endOffset: window.wp.data
+					.select( 'core/block-editor' )
+					.getSelectionEnd().offset,
+			};
+		} );
+	}
+}
+
+test.describe( 'undo', () => {
+	test.beforeEach( async ( { admin } ) => {
+		await admin.createNewPost();
+	} );
+
+	test( 'should undo typing after a pause', async ( {
+		editor,
+		page,
+		pageUtils,
+		undoUtils,
+	} ) => {
+		await page.click( 'role=button[name="Add default block"i]' );
+		await page.keyboard.type( 'before pause' );
+		await editor.page.waitForTimeout( 1000 );
+		await page.keyboard.type( ' after pause' );
+
+		await expect.poll( editor.getBlocks ).toMatchObject( [
+			{
+				name: 'core/paragraph',
+				attributes: { content: 'before pause after pause' },
+			},
+		] );
+
+		await pageUtils.pressKeys( 'primary+z' );
+
+		await expect.poll( editor.getBlocks ).toMatchObject( [
+			{
+				name: 'core/paragraph',
+				attributes: { content: 'before pause' },
+			},
+		] );
+		await expect.poll( undoUtils.getSelection ).toEqual( {
+			blockIndex: 0,
```

Comment:
The value of blockIndex has been reduced by one because the order has been changed to return the order without the title of the post.